### ISSUE TITLE
New compiler: a struct declared in function can be used in other functions

### ISSUE THREAD

**Issue by ivan-mogilko** [adventuregamestudio/ags#1192]
@fernewelten 
With the new compiler you can declare a struct inside a function body. I am not aware if this is intended and a new feature or unintended and a bug.

What's definitely wrong is that you may then use this struct in other functions. What seem at least consistent is that you cannot use it to declare global variables.

Example script:
<pre>
void Test1() {
    struct MyStruct {
        int a;
    }
}

void Test2() {
    MyStruct s;
    Display("%d", s.a);
}
</pre>

Example 2:
<pre>
void Test1() {
    struct MyStruct {
        int a;
    }
}

MyStruct s;
</pre>
This causes Error: 'MyStruct' is illegal outside a function

**Expected behavior**

Whether struct declarations inside a function should be allowed is a topic of it's own (or is this intended feature now?), but definitely a symbol declared locally should be accessible only in the local space it was declared in.

**PS.** On a related note, with the new compiler you can use functions without forward declaring them, but you cannot do same with structs, that is, a struct must be declared strictly above its use. While it's definitely a less problem, I think that's inconsistent.

**Comment by fernewelten** [comment]
Well I've never been _trying_ to allow local `struct` definitions. So from that standpoint it's clearly a bug.
Should we:
- [ ] prevent local struct and enum defs until the compiler is more robust and discuss afterwards whether we want that feature? 
- [ ] leave local struct and enum defs in since this has already happened by chance, but see to it that the feature works properly?

I'm tending to the first option. What do you think?

**Comment by fernewelten** [comment]
There's a good point for making types (structs, enums) valid for the whole program, no matter where they are placed. This might even save some compile time. However, this might break some existing programs, at least in theory. For instance,
```
function oPipe_AnyClick()
{
    int TemperatureGradient = PipeGradient * 0.8; 
    …
}

struct TemperatureGradient
{
    float HotTemperature, LowTemperature;    
    float LiquidDiffusionFactor;
};
TemperatureGradient NuclearReactorGradient;
```

This would work when the struct type `TemperatureGradient` definition is only valid from its point of definition; it would not work if it is valid in the whole program because it would collide with the int variable definition of the same name. 

Question is, how common are programs that are coded like that. At face value I'd say, that's a strange edge case. But perhaps when several pieces of code from different coders are stringed together?

**Comment by ivan-mogilko** [comment]
> * [ ]  prevent local struct and enum defs until the compiler is more robust and discuss afterwards whether we want that feature?
> * [ ]  leave local struct and enum defs in since this has already happened by chance, but see to it that the feature works properly?
> 
> I'm tending to the first option. What do you think?

I agree, it's better to not allow things we don't have clear plan about yet.



> There's a good point for making types (structs, enums) valid for the whole program, no matter where they are placed. This might even save some compile time. However, this might break some existing programs, at least in theory. For instance,
> 
> ```
> function oPipe_AnyClick()
> {
>     int TemperatureGradient = PipeGradient * 0.8; 
>     …
> }
> 
> struct TemperatureGradient
> ```
> 
> This would work when the struct type `TemperatureGradient` definition is only valid from its point of definition; it would not work if it is valid in the whole program because it would collide with the `int` variable definition of the same name.
> 
> Question is, how common are programs that are coded like that. At face value I'd say, that's a strange edge case. But perhaps when several pieces of code from different coders are stringed together?


But is not same case may already take place because of how functions are parsed?
Imo benefit may outweight the chance of this rare problem. A person will have to simply rename the variable.

**Comment by messengerbag** [comment]
If I'm following this idea, I think it might make certain constructs a lot easier to code (though with the limitations against references to custom types in structs, some common patterns still require tedious workarounds), so that's great. At the same time, I have to say I'm quite leery about any suggestion that would add inconsistency to the name scoping in AGS.

But maybe I'm not fully understanding the proposal. Would it imply that a struct or enum type defined inside of a script (rather than in a header) would now be visible inside all other scripts? That seems like a very bad idea. Or are you only talking about making the types valid higher up in the <b>same</b> script? For types declared in headers, would they also be valid in other scripts higher up in the script-order? That would be convenient, but feels quite inconsistent, and might mess with script-modules that users don't understand the internals of.

**Comment by ivan-mogilko** [comment]
@messengerbag the proposal is simply to ignore the order of declarations of types and functions withing one compilation unit.

This will not make the types declared in lower header available in upper script, simply because lower header won't be included when compiling upper scripts.

**Comment by ivan-mogilko** [comment]
Fixed by #1200.
